Index: sys/kern/vfs_subr.c
===================================================================
--- sys/kern/vfs_subr.c	(revision 273960)
+++ sys/kern/vfs_subr.c	(working copy)
@@ -2826,8 +2826,8 @@
 	delmntque(vp);
 	cache_purge(vp);
 	/*
-	 * Done with purge, reset to the standard lock and invalidate
-	 * the vnode.
+	 * Done with purge, notify sleepers of the grim news,
+	 * reset to the standard lock and invalidate  the vnode.
 	 */
 	VI_LOCK(vp);
 	vp->v_vnlock = &vp->v_lock;
@@ -2834,6 +2834,7 @@
 	vp->v_op = &dead_vnodeops;
 	vp->v_tag = "none";
 	vp->v_type = VBAD;
+	VFS_KNOTE_LOCKED(vp, NOTE_REVOKE);
 }
 
 /*
